File: apps/desktop/src-tauri/src/application/levels_service.rs
use std::sync::Arc;

use crate::domain::entities::{LevelConfig, LevelReward, UserLevel};
use crate::domain::ports::LevelRepository;

pub struct LevelsService {
    repo: Arc<dyn LevelRepository>,
}

impl LevelsService {
    pub fn new(repo: Arc<dyn LevelRepository>) -> Self {
        Self { repo }
    }

    pub async fn get_config(&self, guild_id: String) -> Result<LevelConfig, String> {
        self.repo.get_level_config(guild_id).await
    }

    pub async fn get_leaderboard(&self, guild_id: String) -> Result<Vec<UserLevel>, String> {
        self.repo.get_level_leaderboard(guild_id).await
    }

    pub async fn get_rewards(&self, guild_id: String) -> Result<Vec<LevelReward>, String> {
        self.repo.get_level_rewards(guild_id).await
    }

    pub async fn set_reward(&self, guild_id: String, level: i32, role_id: String, source: String) -> Result<LevelReward, String> {
        self.repo.set_level_reward(guild_id, level, role_id, source).await
    }

    pub async fn delete_reward(&self, guild_id: String, level: i32, source: String) -> Result<(), String> {
        self.repo.delete_level_reward(guild_id, level, source).await
    }
}

File: apps/web/src/services/levelsService.ts
import { httpGet, httpPost, httpDelete } from "@/api/http";
import type { LevelConfig, UserLevel, LevelReward } from "@/types";
import { q } from "./_query";

export interface SaveLevelConfigPayload {
  guild_id: string;
  xp_per_message?: number;
  xp_per_voice_minute?: number;
  xp_cooldown_secs?: number;
  level_up_channel_id?: string | null;
  level_up_message?: string;
  excluded_channels?: string[];
  enabled?: boolean;
}

export interface AddXpPayload {
  guild_id: string;
  user_id: string;
  username: string;
  amount: number;
  source?: "text" | "voice";
}

export const levelsService = {
  getConfig(guildId: string): Promise<LevelConfig> {
    return httpGet(`/api/levels/config/${guildId}`);
  },
  saveConfig(body: SaveLevelConfigPayload): Promise<LevelConfig> {
    return httpPost(`/api/levels/config`, body);
  },
  addXp(body: AddXpPayload): Promise<unknown> {
    return httpPost(`/api/levels/xp`, body);
  },
  getLeaderboard(guildId: string): Promise<UserLevel[]> {
    return httpGet(`/api/levels/${guildId}/leaderboard`);
  },
  getRewards(guildId: string): Promise<LevelReward[]> {
    return httpGet(`/api/levels/rewards/${guildId}`);
  },
  setReward(guildId: string, level: number, roleId: string, source: string): Promise<unknown> {
    return httpPost("/api/levels/rewards", { guild_id: guildId, level, role_id: roleId, source });
  },
  deleteReward(guildId: string, level: number, source: string): Promise<unknown> {
    return httpDelete(`/api/levels/rewards/${guildId}/${level}${q({ source })}`);
  },
};
